File: GroceryListHelper.Client/Services/CartProductsLocalService.cs
namespace GroceryListHelper.Client.Services;

public sealed class CartProductsLocalService(ILocalStorageService localStorage, AppState appState) : ICartProductsService
{
    private const string cartProductsKey = "cartProducts";

    public async Task<List<CartProductCollectable>> GetCartProducts()
    {
        List<CartProductCollectable> products = await localStorage.GetItemAsync<List<CartProductCollectable>>(cartProductsKey) ?? [];
        return products;
    }

    public async Task CreateCartProduct(CartProduct product)
    {
        await localStorage.SetItemAsync(cartProductsKey, appState.CartProducts);
    }

    public async Task DeleteCartProduct(string name)
    {
        await localStorage.SetItemAsync(cartProductsKey, appState.CartProducts);
    }

    public async Task DeleteAllCartProducts()
    {
        await localStorage.RemoveItemAsync(cartProductsKey);
    }

    public async Task UpdateCartProduct(CartProductCollectable cartProduct)
    {
        await localStorage.SetItemAsync(cartProductsKey, appState.CartProducts);
    }

    public async Task SortCartProducts(ListSortDirection sortDirection)
    {
        await localStorage.SetItemAsync(cartProductsKey, appState.CartProducts);
    }
}

File: GroceryListHelper.Client/src/services/CartProductsLocalService.ts
import type { CartProduct } from "../types/CartProduct";
import type { ICartProductsService } from "../types/ICartProductsService";
import type { SortDirection } from "../types/SortState";
import { cartProductsKey } from "../helpers/globalConstants";
import store from "../helpers/store.svelte";

export class CartProductsLocalService implements ICartProductsService {
    getCartProducts = async () => {
        try {
            const productsString = localStorage.getItem(cartProductsKey);
            const products = productsString ? JSON.parse(productsString) as CartProduct[] : [];
            return products;
        }
        catch (e) {
            if (e instanceof Error) {
                return e;
            }
            else {
                return new Error("Unknow problem loading from local storage");
            }
        }
    }

    deleteAllCartProducts = () => {
        return this.saveCartProductsToLocalStorage();
    }

    createCartProduct = (product: CartProduct) => {
        return this.saveCartProductsToLocalStorage();
    }

    updateCartProduct = (product: CartProduct) => {
        return this.saveCartProductsToLocalStorage();
    }

    sortCartProducts = (sortDirection: SortDirection) => {
        return this.saveCartProductsToLocalStorage();
    }

    deleteCartProduct = (name: string) => {
        return this.saveCartProductsToLocalStorage();
    }

    private async saveCartProductsToLocalStorage() {
        try {
            localStorage.setItem(cartProductsKey, JSON.stringify(store.cartProducts));
            return null;
        }
        catch (e) {
            if (e instanceof Error) {
                return e;
            }
            else {
                return new Error("Unknow problem saving to local storage");
            }
        }
    }
}